graylog2-web-interface/src/components/security/investigations/AddEvidence.test.tsx

```diff
@@ -0,0 +1,193 @@
+
+type SecEvidenceType = {
+  invId: string,
+  payload: {
+    dashboards?: string[],
+    searches?: string[],
+    events?: string[],
+    logs?: { id: string, index: string }[],
+  },
+};
+
+const mockSecAddEvidence = jest.fn((_payload: SecEvidenceType) => {});
+const mockUseAddEvidence = jest.fn(() => ({
+  addEvidence: mockSecAddEvidence,
+  addingEvidence: false,
+}));
+
+const mockUseInvestigationDrawer = jest.fn(() => ({
+  selectedInvestigationId: 'investigation-id',
+}));
+
+type SecAddEvidenceProps = {
+  id: string,
+  type: string,
+  children: React.ReactNode,
+  index?: string,
+};
+
+const MockSecAddEvidence = ({ id, index, type, children }: SecAddEvidenceProps) => {
+  const { addEvidence } = mockUseAddEvidence();
+  const { selectedInvestigationId } = mockUseInvestigationDrawer();
+
+  const submitEvidence = (e: React.BaseSyntheticEvent) => {
+    e.preventDefault();
+    e.stopPropagation();
+
+    switch (type) {
+      case 'logs':
+        addEvidence({
+          invId: selectedInvestigationId,
+          payload: { logs: [{ id, index }] },
+        });
+
+        break;
+      case 'dashboards':
+      case 'searches':
+      case 'events':
+        addEvidence({
+          invId: selectedInvestigationId,
+          payload: { [type]: [id] },
+        });
+
+        break;
+      default:
+        break;
+    }
+  };
+
+  const childrenWithProps = React.Children.map(children, (child: React.ReactElement) => (
+    React.cloneElement(child, { onClick: submitEvidence })
+  ) as React.ReactElement);
+
+  // eslint-disable-next-line react/jsx-no-useless-fragment
+  return <>{childrenWithProps}</>;
+};
+
+MockSecAddEvidence.defaultProps = {
+  index: undefined,
+};
+
+const mockExports = jest.fn((_entityKey?: string) => ([{
+  components: {
+    AddEvidence: MockSecAddEvidence,
+  },
+  hooks: {
+    useInvestigationDrawer: mockUseInvestigationDrawer,
+  },
+}]));
+
+jest.mock('graylog-web-plugin/plugin', () => ({
+  PluginStore: { exports: jest.fn(() => mockExports()) },
+}));
+
+const AddToInvestigation = ({ id, type, index }: { id: string, type: EvidenceTypes, index?: string }) => {
+  const [disableButton, setDisableButton] = React.useState<boolean>(false);
+
+  return (
+    <AddEvidence id={id}
+                 type={type}
+                 index={index}
+                 investigationSelected={(arg: boolean) => setTimeout(() => setDisableButton(!arg), 10)}>
+      <button type="button" disabled={disableButton}>Add to investigation</button>
+    </AddEvidence>
+  );
+};
+
+AddToInvestigation.defaultProps = {
+  index: undefined,
+};
+
+describe('AddEvidence', () => {
+  it('should render button', () => {
+    render(<AddToInvestigation id="dashboard-id" type="dashboards" />);
+
+    expect(screen.getByText('Add to investigation')).toBeInTheDocument();
+  });
+
+  it('should not fail and render null if no plugin is registered', () => {
+    mockExports.mockReturnValueOnce(undefined);
```

Comment:
Since we use `mockReturnValueOnce` it does not really reflect the actual behaviour. I personally prefer the following approach:

(Optionally) Not defining any return value in the initial mock
```
jest.mock('graylog-web-plugin/plugin', () => ({
  PluginStore: { exports: jest.fn() },
}));
```

Using `beforeEach` instead to mock the return value:
```
  beforeEach(() => {
    asMock(PluginStore.exports).mockReturnValue(mockExports());
  });
```

In the test we can now use `mockReturnValue`:
```
asMock(PluginStore.exports).mockReturnValue(undefined);
```

In this test the benefit is neglectable, but in general this has the benefits,
- That mocking in a test never affects the following tests, because we always reset the mock.
- We can use `mockReturnValue` and can be sure our test specific mock is consistent, even after a rerender.
- It is slightly easier to read, since you do not need to look up what module e.g. `mockUseInvestigationDrawer` is relating to.

The mocking could still be simple if we use default values for `mockExports`:
```
const mockExports = (AddEvidence = MockSecAddEvidence, useInvestigationDrawer = mockUseInvestigationDrawer) =>

...
asMock(PluginStore.exports).mockReturnValue(mockExports(undefined, { selectedInvestigationId: undefined });
```

Reply:
In this case, I'll have to disagree. The intention is to test the plugin functionality and how the proxy element handles the absence of the plugin.

So the absence of the plugin is the exception. In my mind, it makes more sense to have one case that uses `mockReturnValueOnce` to test the exception and for the rest of the tests, use the default behavior with the plugin present.

Reply:
I think we have the same idea of the general process.

In my suggestion the plugin would be present by default, since it will always be defined for each test with `beforeEach`. 

I just did not defined it in the initial `jest.mock`, since this will always be overwritten by `beforeEach`.
In my suggestion the `jest.mock` is only necessary to be able to mock the module.

My main point is, when we use `mockReturnValueOnce`, `PluginStore.exports` will only be `undefined` (in this test case) on the first render. When the component rerenders we call `PluginStore.exports` again and the plugin will be present again, although this would not be the case in production.

In every case this is just a general pattern I wanted to talk about and it is neglectable for this case and I have no problem with merging the current state.

Reply:
I just saw that we are using `usePluginEntities`, which memorises the return value of `PluginStore.exports`, which means my suggestion ist even less relevant for this specific case.